### PR TITLE
feat(components): [carousel] height is adaptive when height is auto

### LINKED ISSUE #11368: [Component] [carousel, carousel-item] Carousel 走马灯 不主动设置 height 属性高度  Carousel Item 无法自适应插槽的高度
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.45`
- Element Plus Version: `2.2.28`
- Browser / OS: `UserAgent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-carousel`
- `el-carousel-item`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtY2Fyb3VzZWwgOmludGVydmFsPVwiNDAwMFwiID5cbiAgICA8ZWwtY2Fyb3VzZWwtaXRlbSB2LWZvcj1cIml0ZW0gaW4gNlwiIDprZXk9XCJpdGVtXCI+XG4gICAgICA8aDMgdGV4dD1cIjJ4bFwiIGp1c3RpZnk9XCJjZW50ZXJcIj57eyBpdGVtIH19PC9oMz5cbiAgICA8L2VsLWNhcm91c2VsLWl0ZW0+XG4gIDwvZWwtY2Fyb3VzZWw+XG48L3RlbXBsYXRlPlxuXG48c3R5bGUgc2NvcGVkPlxuLmVsLWNhcm91c2VsX19pdGVtIGgzIHtcbiAgY29sb3I6ICM0NzU2Njk7XG4gIG9wYWNpdHk6IDAuNzU7XG4gIGxpbmUtaGVpZ2h0OiAyMDBweDtcbiAgbWFyZ2luOiAwO1xuICB0ZXh0LWFsaWduOiBjZW50ZXI7XG59XG5cbi5lbC1jYXJvdXNlbF9faXRlbTpudGgtY2hpbGQoMm4pIHtcbiAgYmFja2dyb3VuZC1jb2xvcjogIzk5YTliZjtcbiAgaGVpZ2h0OiA1MDBweFxufVxuXG4uZWwtY2Fyb3VzZWxfX2l0ZW06bnRoLWNoaWxkKDJuICsgMSkge1xuICBiYWNrZ3JvdW5kLWNvbG9yOiAjZDNkY2U2O1xufVxuPC9zdHlsZT5cbiIsImltcG9ydF9tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwiX28iOnt9fQ==)

### Steps to reproduce

不主动设置 height

### What is Expected?

看图

### What is actually happening?

看图

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by qq282126990**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Review by tolking — APPROVED** [review]

File: packages/theme-chalk/src/menu.scss [hunk 1]
```diff
@@ -167,7 +167,10 @@
 
     > .#{$namespace}-menu-item,
     > .#{$namespace}-sub-menu > .#{$namespace}-sub-menu__title,
-    > .#{$namespace}-menu-item-group > ul > .#{$namespace}-sub-menu > .#{$namespace}-sub-menu__title {
+    > .#{$namespace}-menu-item-group
+      > ul
+      > .#{$namespace}-sub-menu
+      > .#{$namespace}-sub-menu__title {
```
Comment (btea) [inline packages/theme-chalk/src/menu.scss:173]: Doesn't seem to be modified here?

Comment (qq282126990) [reply packages/theme-chalk/src/menu.scss:173]: This is generated after formatting pnpm format

Comment (qq282126990) [reply packages/theme-chalk/src/menu.scss:173]: I resubmit

Comment (qq282126990) [reply packages/theme-chalk/src/menu.scss]: has been resubmitted @btea

Comment (qq282126990) [reply packages/theme-chalk/src/menu.scss]: I feel that there are still several files after pnpm format. There will be formatted files. I suggest you reformat them once.

Comment (btea) [reply packages/theme-chalk/src/menu.scss]: Ok, I see.

**Comment by btea** [comment]
Hi, I took a look at the example you provided. IMO, this feature seems unnecessary,  if used in the project like this, it will cause the content of the page to shake, which is very strange. Can you explain why this functionality needs to be added?

**Comment by qq282126990** [comment]
> 你好，我看了你提供的例子。IMO，这个功能似乎是不必要的，如果在这样的项目中使用，会导致页面内容抖动，这是很奇怪的。您能解释一下为什么需要添加此功能吗？

My idea is that the user can control the height of the carousel by controlling the carousel-item, not just by setting the height

**Comment by qq282126990** [comment]
https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPHA+6Ieq5a6a5LmJIGl0ZW0g6auY5bqmPC9wPlxuICA8ZWwtY2Fyb3VzZWwgaGVpZ2h0PVwiYXV0b1wiIDphdXRvcGxheT1cImZhbHNlXCI+XG4gICAgPGVsLWNhcm91c2VsLWl0ZW0gc3R5bGU9XCJoZWlnaHQ6IDEwMHB4XCI+XG4gICAgICA8aDMgY2xhc3M9XCJzbWFsbCBqdXN0aWZ5LWNlbnRlclwiIHRleHQ9XCIyeGxcIj5oZWlnaHQgMTAwcHg8L2gzPlxuICAgIDwvZWwtY2Fyb3VzZWwtaXRlbT5cbiAgICA8ZWwtY2Fyb3VzZWwtaXRlbSBzdHlsZT1cImhlaWdodDogMTAwcHhcIj5cbiAgICAgIDxoMyBjbGFzcz1cInNtYWxsIGp1c3RpZnktY2VudGVyXCIgdGV4dD1cIjJ4bFwiPmhlaWdodCAxMDBweDwvaDM+XG4gICAgPC9lbC1jYXJvdXNlbC1pdGVtPlxuICAgIDxlbC1jYXJvdXNlbC1pdGVtIHN0eWxlPVwiaGVpZ2h0OiAxMDBweFwiPlxuICAgICAgPGgzIGNsYXNzPVwic21hbGwganVzdGlmeS1jZW50ZXJcIiB0ZXh0PVwiMnhsXCI+aGVpZ2h0IDEwMHB4PC9oMz5cbiAgICA8L2VsLWNhcm91c2VsLWl0ZW0+XG4gIDwvZWwtY2Fyb3VzZWw+XG4gIDxwPuiuvue9riBoZWlnaHQ9MzAwIDwvcD5cbiAgPGVsLWNhcm91c2VsIGhlaWdodD1cIjMwMFwiIDphdXRvcGxheT1cImZhbHNlXCI+XG4gICAgPGVsLWNhcm91c2VsLWl0ZW0gPlxuICAgICAgPGgzIGNsYXNzPVwic21hbGwganVzdGlmeS1jZW50ZXJcIiB0ZXh0PVwiMnhsXCI+aGVpZ2h0IDEwMHB4PC9oMz5cbiAgICA8L2VsLWNhcm91c2VsLWl0ZW0+XG4gICAgPGVsLWNhcm91c2VsLWl0ZW0gPlxuICAgICAgPGgzIGNsYXNzPVwic21hbGwganVzdGlmeS1jZW50ZXJcIiB0ZXh0PVwiMnhsXCI+aGVpZ2h0IDEwMHB4PC9oMz5cbiAgICA8L2VsLWNhcm91c2VsLWl0ZW0+XG4gICAgPGVsLWNhcm91c2VsLWl0ZW0+XG4gICAgICA8aDMgY2xhc3M9XCJzbWFsbCBqdXN0aWZ5LWNlbnRlclwiIHRleHQ9XCIyeGxcIj5oZWlnaHQgMTAwcHg8L2gzPlxuICAgIDwvZWwtY2Fyb3VzZWwtaXRlbT5cbiAgPC9lbC1jYXJvdXNlbD5cbjwvdGVtcGxhdGU+XG5cbjxzdHlsZSBzY29wZWQ+XG4uY2Fyb3VzZWwtaXRlbSB7XG4gIGNvbG9yOiAjNDc1NjY5O1xuICBvcGFjaXR5OiAwLjc1O1xuICBtYXJnaW46IDA7XG4gIHRleHQtYWxpZ246IGNlbnRlcjtcbn1cbi5lbC1jYXJvdXNlbF9faXRlbSBoMyB7XG4gIGNvbG9yOiAjNDc1NjY5O1xuICBvcGFjaXR5OiAwLjc1O1xuICBkaXNwbGF5OiBmbGV4O1xuICBhbGlnbi1pdGVtczogY2VudGVyO1xuICBtYXJnaW46IDA7XG4gIHRleHQtYWxpZ246IGNlbnRlcjtcbiAgaGVpZ2h0OiAxMDAlO1xufVxuLmVsLWNhcm91c2VsX19pdGVtOm50aC1jaGlsZCgybikge1xuICBiYWNrZ3JvdW5kLWNvbG9yOiAjOTlhOWJmO1xufVxuLmVsLWNhcm91c2VsX19pdGVtOm50aC1jaGlsZCgybiArIDEpIHtcbiAgYmFja2dyb3VuZC1jb2xvcjogI2QzZGNlNjtcbn1cbjwvc3R5bGU+IiwiUGxheWdyb3VuZE1haW4udnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgc2V0dXBFbGVtZW50UGx1cyB9IGZyb20gJy4vZWxlbWVudC1wbHVzLmpzJ1xuc2V0dXBFbGVtZW50UGx1cygpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0X21hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9wcmV2aWV3LTExMzkyLWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwidW5zdXBwb3J0ZWRcIlxuICB9XG59IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AdnVlL3J1bnRpbWUtZG9tQGxhdGVzdC9kaXN0L3J1bnRpbWUtZG9tLmVzbS1icm93c2VyLmpzXCIsXG4gICAgXCJAdnVlL3NoYXJlZFwiOiBcImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9zaGFyZWRAbGF0ZXN0L2Rpc3Qvc2hhcmVkLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctMTEzOTItZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcbmltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgIGxpbmsuaHJlZiA9ICdodHRwczovL3ByZXZpZXctMTEzOTItZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5jc3MnXG4gICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdsb2FkJywgcmVzb2x2ZSlcbiAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgIGRvY3VtZW50LmJvZHkuYXBwZW5kKGxpbmspXG4gIH0pXG59IiwiX28iOnsic2hvd0hpZGRlbiI6dHJ1ZSwic3R5bGVTb3VyY2UiOiJodHRwczovL3ByZXZpZXctMTEzOTItZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5jc3MifX0=

**Comment by qq282126990** [comment]
> Hi, I took a look at the example you provided. IMO, this feature seems unnecessary, if used in the project like this, it will cause the content of the page to shake, which is very strange. Can you explain why this functionality needs to be added?

Do you think it is necessary to add this feature

File: packages/components/carousel/src/use-carousel.ts [hunk 2]
```diff
@@ -192,6 +204,12 @@ export const useCarousel = (
     startTimer()
   }
 
+  function setContainerHeight(height: number) {
+    if (props.height === 'auto') {
+      containerHeight.value = height
+    }
+  }
```
Comment (jw-foss) [inline packages/components/carousel/src/use-carousel.ts:211]: **Non-blocking:** For my personal preference, I would write this in this way:

```suggestion
  function setContainerHeight(height: number) {
    if (props.height !== 'auto') return
    containerHeight.value = height
  }
```

For me early return would make the code more readable.

**Review by jw-foss — COMMENTED** [review]
@qq282126990 thank you for contributing, the code LGTM, I have left a couple comments for your consideration. And I'd like to ask you to add a test case against your changes. Let me know if you have any questions 👍

**Comment by jw-foss** [comment]
Also, in addition to @btea's question, maybe we could use a flag to control whether we should enable auto height for the users. Just to make sure the changes **WILL NOT** cause any trouble.

File: packages/components/carousel/src/use-carousel.ts (hunk 2 above)
Comment (qq282126990) [reply packages/components/carousel/src/use-carousel.ts:211]: Thank you, let me modify it

**Comment by qq282126990** [comment]
> 

I think the height can be set to auto by default, and the user needs to manually set it when the user needs to set the height

**Comment by jw-foss** [comment]
> > 
> 
> I think the height can be set to auto by default, and the user needs to manually set it when the user needs to set the height

Just so you know that there were users who use this component and the changes you made **might** change the default behavior for those users per @btea's question above. Marking this behind a flag would be more flexible.

**Comment by qq282126990** [comment]
> 

ok, I feel that I need to add a new attribute named auto-height to delete and modify height to 'auto'

**Comment by qq282126990** [comment]
> 我认为没有必要添加新的属性，默认值`height`仍然是`''`。本次 PR 没有 BC，不会影响之前的用户

So should I keep my original judgment?

**Comment by qq282126990** [comment]
@jw-foss Using the test case, I don’t know why the height of the element cannot be obtained

image
image

**Comment by qq282126990** [comment]
In addition to avoid the first rendering height of 0 should need to add a nextTick

image

**Comment by qq282126990** [comment]
> 

image

**Comment by qq282126990** [comment]
> > 
> 
> 图像

image

**Comment by qq282126990** [comment]
> @qq282126990感谢您贡献代码 LGTM，我留下了一些评论供您考虑。我想请您针对您的更改添加一个测试用例。如果您有任何问题，请告诉我👍

hello, Here is the test case I wrote

image

One problem is that running the test case carousel-item cannot get offsetHeight


image

**Comment by Flourad** [comment]
这个特性可用了么？

**Comment by btea** [comment]
> hello, Here is the test case I wrote
> 
> image
> 
> One problem is that running the test case carousel-item cannot get offsetHeight
> 
> image
> 
This seems to be a jsdom bug, refer [here](https://github.com/jsdom/jsdom/issues/135).

**Comment by qq282126990** [comment]
ok ,i see

**Comment by btea** [comment]
@qq282126990 There seems to be a conflict in the code, can you fix it?

**Comment by qq282126990** [comment]
ok

**Comment by qq282126990** [comment]
Hello, there may be problems with the previous branch, I recreated a new branch, please check the pull request

https://github.com/element-plus/element-plus/pull/12388

@btea

**Pull request closed without merging**